### ISSUE TITLE
Patching should convert datetime strings

### ISSUE THREAD

**Issue by dereuromark** [cakephp/cakephp#10586]
This is a (multiple allowed):

* [] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.4.5
* MySQL 5.7

### What you did
A basic JSON export and then import on a different app/DB:
```
                $json = json_decode($content, true);
                ...
                foreach ($json as $row) {
                    unset($row['id']);
                    //$row['created'] is '2017-04-05T17:18:00+00:00',
                    //$row['modified'] is '2017-04-05T17:18:00+00:00',
                    $release = $this->Releases->newEntity($row);

                    $this->Releases->saveOrFail($release);
                }
```

### What happened
```
[PDOException] SQLSTATE[22007]: Invalid datetime format: 
1292 Incorrect datetime value: '2017-04-05T17:18:00+00:00' for column 'created' at row 1
```

After marshalling the JSON ISO format strings are still strings not have not been converted to Time() objects that the ORM can safely handle for saving.

### What you expected to happen
Saving just fine - including the JSON datetime string.

Maybe the patching should check the fields based on the schema, and similar to array=>object for date/time, should also transform ISO strings to Time() objects?
That would avoid manual transformation or unnecessary exceptions on save.

Any takes on where this would best be done?
I can then take a stab at it when I find some time.

**Comment by ADmad** [comment]
The `DateTimeType` db type class already handles converting strings to objects when marshalling, if the datetime string is in the format the type class is configured for. If you haven't turned on locale based parsing then the default format is `Y-m-d H:i:s`.

**Comment by dereuromark** [comment]
I dont think the JSON ISO code counts as localized - it is as valid as Y-m-d H:i:s directly IMO.

**Comment by dereuromark** [comment]
I think prematurely setting tickets on hold is also not too nice - I must say.
You should at least wait for some more feedback before doing so.

Could the _format be an array of acceptable ISO standard formats?
```php
    /**
     * String format to use for DateTime parsing
     *
     * @var string|array
     */
    protected $_format = [
        'Y-m-d H:i:s',
        ...
    ];
```

Refs https://en.wikipedia.org/wiki/ISO_8601

**Comment by dereuromark** [comment]
Another idea:
```php
// Normalize ISO code
if (preg_match('^/[\d]{2}-[\d]{2}-[\d]{2}T[\d]{2}:[\d]{2}:[\d]{2}[+-][\d]{2}:[\d]{2}$/') {
     // Normalize into expected internal ISO format incl the timezone shift
}
```

**Comment by ADmad** [comment]
> I dont think the JSON ISO format counts as localized - it is as valid as Y-m-d H:i:s directly IMO.

Well `2017-04-05T17:18:00+00:00` is not in `Y-m-d H:i:s` format.

```phpPsy Shell v0.8.2 (PHP 7.1.2-4+deb.sury.org~xenial+1 — cli) by Justin Hileman
New version is available (current: v0.8.2, latest: v0.8.3)
>>> \Cake\I18n\Time::parseDateTime('2017-04-05T17:18:00+00:00', 'Y-m-d H:i:s')
=> null
>>> \Cake\I18n\Time::parseDateTime('2017-04-05 17:18:00', 'Y-m-d H:i:s')
=> Cake\I18n\Time {#252
     +"time": "2017-01-05T17:04:00+00:00",
     +"timezone": "UTC",
     +"fixedNowTime": false,
   }
```

> You should at least wait for some more feedback before doing so.

Your title says "Patching should convert datetime strings"  which is already done (if it's in proper format), hence the "on hold" to get more feedback. Waiting for feedback/action from the poster is the case "On hold" label is for.

Only after my response did you come up with the enhancement idea of converting `_format` to an array instead of string.

**Comment by dereuromark** [comment]
This could almost be marked as a bugfix even, your example shows the defect here rather obvious:

    "time": "2017-01-05T17:04:00+00:00"

is the internal value (normalized), but the same string as input would not be accepted, which is just wrong.
We should definitely fix this.

**Comment by markstory** [comment]
The internals of the `DateTimeType` currently do

```php
        try {
            $compare = $date = false;
            if ($value === '' || $value === null || $value === false || $value === true) {
                return null;
            }
            if (is_numeric($value)) {
                $date = new $class('@' . $value);
            } elseif (is_string($value) && $this->_useLocaleParser) {
                return $this->_parseValue($value);
            } elseif (is_string($value)) {
                $date = new $class($value);
                $compare = true;
            }
            if ($compare && $date && $date->format($this->_format) !== $value) {
                return $value;
            }
            if ($date) {
                return $date;
            }
        } catch (Exception $e) {
            return $value;
        }
```

I'm guessing that `new FrozenDateTime($value)` doesn't work with ISO formatted strings.

**Comment by dereuromark** [comment]
@markstory The problem is that 

    && $date->format($this->_format) !== $value

will never work this way with strings coming in including the T and/or the +-MM:SS part.
So either normalizing beforehand or comparing in an array of formats would resolve it.

**Comment by markstory** [comment]
Enabling multiple datetime formats could be useful. But I wouldn't want to ship with multiple default formats.

**Comment by markstory** [comment]
Closing as there is a pull request.

**Comment by yark** [comment]
Interesting. What i'm missing here?

    >>> \Cake\I18n\Time::parseDateTime('2017-04-05 00:00:00', 'Y-m-d H:i:s')
    => null

**Comment by ADmad** [comment]
The format string uses ICU date time placeholders not `strtotime()` style place holders.
https://api.cakephp.org/3.5/class-Cake.I18n.DateFormatTrait.html#_parseDateTime
http://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table

**Comment by yark** [comment]
@ADmad i have copied that sample from your comment in this tread from 26 Apr 2017
It was changed for this year?
